File: hw2/code.cpp
#include <chrono>
#include <thread>
#include <iostream>
#include <vector>

using namespace std;
using namespace std::this_thread;    
using namespace std::chrono_literals; 
using std::chrono::system_clock;

void timeconsuming()
{
    sleep_for(5s);
}

void for_loop(int n)
{
    for (int i = 0; i < n; i++)
    {
        timeconsuming();
    }
}

void worker_threads(int n)
{
    std::vector<thread> threads(4);
    for (int i = 0; i < 4; i++)
    {
        threads[i] = thread(timeconsuming);
    }
    for (auto &th : threads)
    {
        th.join();
    }
}
int main()
{

    int n = 20;
    auto start = chrono::steady_clock::now();
    for_loop(n);
    auto end = chrono::steady_clock::now();
    auto sequential_time = end - start;

    start = chrono::steady_clock::now();
    worker_threads(n);
    end = chrono::steady_clock::now();
    auto parallel_time = end - start;

    float speedup = sequential_time/parallel_time;

    cout << "Sequential Elapsed time: " << std::chrono::duration_cast<std::chrono::seconds>(sequential_time).count() << endl;
    cout << "Parallel Elapsed time: " << std::chrono::duration_cast<std::chrono::seconds>(parallel_time).count() << endl;
    cout << "Speed Up: " << speedup << endl;
    cout << "Average Efficiency: " << speedup/4  << endl;
}

File: hw2/code.py
import time
import threading
import queue
import random
import multiprocessing

def timeconsuming():
    time.sleep(random.randint(1, 5))
    return

n = 40

start = time.time()
for i in range(0, n):
    timeconsuming()
end = time.time()
sequential_time = end - start
print("Sequential Time Elapsed: " , sequential_time)

q = queue.Queue()
for i in range(n):
    q.put(i)

def worker():
    while True:
        timeconsuming()
        q.task_done()

cpus = multiprocessing.cpu_count()
print("Creating %d threads" % cpus)
start = time.time()
for i in range(cpus):
    t = threading.Thread(target=worker)
    t.daemon = True
    t.start()

q.join()
end = time.time()

parallel_time = end - start

print("Parallel Time Elapsed: " , parallel_time)

# Stats
speedup = sequential_time/parallel_time
avg_efficiency = speedup/4
print("SpeedUp:" , speedup)
print("Average Efficiency:" , avg_efficiency)